src/hotspot/share/memory/universe.cpp

```diff
@@ -924,55 +923,69 @@ void universe_oopstorage_init() {
   Universe::oopstorage_init();
 }
 
-static void initialize_known_method(LatestMethodCache* method_cache,
-                                    InstanceKlass* ik,
-                                    const char* method,
-                                    Symbol* signature,
-                                    bool is_static, TRAPS)
+void LatestMethodCache::init(JavaThread* current, InstanceKlass* ik,
+                             const char* method, Symbol* signature, bool is_static)
 {
   TempNewSymbol name = SymbolTable::new_symbol(method);
   Method* m = nullptr;
   // The klass must be linked before looking up the method.
-  if (!ik->link_class_or_fail(THREAD) ||
+  if (!ik->link_class_or_fail(current) ||
       ((m = ik->find_method(name, signature)) == nullptr) ||
       is_static != m->is_static()) {
-    ResourceMark rm(THREAD);
+    ResourceMark rm(current);
     // NoSuchMethodException doesn't actually work because it tries to run the
     // <init> function before java_lang_Class is linked. Print error and exit.
     vm_exit_during_initialization(err_msg("Unable to link/verify %s.%s method",
                                  ik->name()->as_C_string(), method));
   }
-  method_cache->init(ik, m);
+
+  _klass = ik;
+  _method_idnum = m->method_idnum();
+  assert(_method_idnum >= 0, "sanity check");
+}
+
+Method* LatestMethodCache::get_method() {
+  if (klass() == nullptr) return nullptr;
+  InstanceKlass* ik = InstanceKlass::cast(klass());
```

Comment:
Sorry for the bits and pieces.  Hopefully the last thing.  Can _klass be a InstanceKlass so we don't need this cast?

Reply:
Thanks for the suggestions. This code is very old and really needs to be cleaned up. I also removed the unnecessary accessor functions.